### PR TITLE
Mod4.0 - User Profile & Following

### PR DESCRIPTION
**Pull request by ppj**

### CODE REVIEW

File: app/models/following.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  belongs_to :followed, class_name: "User"
+  validates_presence_of :follower_id
+  validates_presence_of :followed_id
+  validates_uniqueness_of :follower_id, scope: :followed_id
+  validate :cannot_follow_self
+
+  private
+
+  def cannot_follow_self
+    @errors.add(:base, "You cannot follow yourself") if follower_id == followed_id
```
Comment (brandoncc) [inline app/models/following.rb:12]: Nice use of a custom validation :smile:

File: app/models/user.rb [hunk 2]
```diff
@@ -16,4 +22,12 @@ def normalize_queue_item_positions
   def queued?(video)
     queue_items.map(&:video).include? video
```
Comment (brandoncc) [inline app/models/user.rb]: Just so you know, a more performant way to do this would be:

``` ruby
queue_items.pluck(:video_id).include? video.id
```

Pluck pulls only the ids from the database and loads the data into an array. It saves a lot of processing time and power over loading every video into memory using `map`.

File: spec/controllers/followings_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
+require 'spec_helper'
+
+describe FollowingsController do
+  describe "GET index" do
+    it_behaves_like "a security guard" do
```
Comment (brandoncc) [inline spec/controllers/followings_controller_spec.rb]: If you create a lot of shared example scenarios, you are going to want the descriptions to be accurate. What is the security guard doing? Is he not letting anybody in? Is he only letting in admins? Or anybody that is logged in? Your specs should be very easy to read, this one is not unfortunately.

File: spec/controllers/followings_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      get :index
+      expect(assigns(:followings)).to eq([following])
+    end
+  end
+
+  describe "DELETE destroy" do
+    let(:bob) { Fabricate :user }
+    let(:jane) { Fabricate :user }
+    let(:following) { following = Fabricate :following, follower: bob, followed: jane }
+    before { set_current_user bob }
```
Comment (brandoncc) [inline spec/controllers/followings_controller_spec.rb]: Be careful with `let` and `before` I find your specs in this group very hard to follow because you have to keep jumping up to this section to read them. I would recommend setting up your specs in each `it` as much as possible.

File: app/models/following.rb (hunk 1 above)
Comment (ppj) [inline app/models/following.rb:12]: Thanks :smiley:
I was keen on pushing this validation to the model than keeping it in the controller. I thought this was the simplest way.

File: app/models/user.rb (hunk 2 above)
Comment (ppj) [inline app/models/user.rb]: Good to know this Brandon. Thanks!

File: spec/controllers/followings_controller_spec.rb (hunk 4 above)
Comment (ppj) [inline spec/controllers/followings_controller_spec.rb]: Agreed. I made the same mistake Kevin warns about in the lessons. Fixed most such occurrences where the reader would have to scroll too much to see the "let"s and "before"s. Thanks.

File: spec/controllers/followings_controller_spec.rb (hunk 3 above)
Comment (ppj) [inline spec/controllers/followings_controller_spec.rb]: I agree that it doesn't read well in the code, but I named it so because it reads better when run the spec in documentation mode.

```
UsersController
  GET show
    behaves like a security guard
      redirects to the root page
```

I renamed it "a gatekeeper redirecting an unauthenticated user" :smiley:. What do you think?

Comment (brandoncc) [inline spec/controllers/followings_controller_spec.rb]: That sounds good :-)
